### PR TITLE
Messaging sockets

### LINKED ISSUE #28: IN: Messaging sockets
Integrate sockets into messaging so that messages are sent instantly when users are online

### PR DESCRIPTION
**Pull request by g4rry420**
### What this PR does (required):
- Added messaging sockets for the messaging page

### Screenshots / Videos (required):
N/A

### Any information needed to test this feature (required):
- Go to `messages` page and chat with user from another browser.

### Any issues with the current functionality (optional):
No

### CODE REVIEW

File: client/src/pages/Messages/ActiveChat/Input.tsx [hunk 1]
```diff
@@ -67,24 +69,66 @@ const useStyles = makeStyles((theme) => ({
...
+    if (socket === undefined || currentUserId === '') return;
+
+    socket.on('new-message', (data) => {
+      dispatchMessages({ type: 'ADD_NEW_MESSAGE', activeConversation: data.activeConversation, message: data.message });
+      dispatchConversations({
+        type: 'UPDATE_LATEST_MESSAGE',
+        activeConversation: data.activeConversation,
+        message: data.message.text,
+        createdAt: new Date().toISOString(),
+      });
```
Comment (dproc96) [inline client/src/pages/Messages/ActiveChat/Input.tsx]: (optional) These two actions always fire together, should they be combined?

File: client/src/pages/Messages/index.tsx [hunk 2]
```diff
@@ -50,6 +52,23 @@ export default function Messages(): JSX.Element {
...
 
+  useEffect(() => {
+    if (socket === undefined) return;
+
+    socket.on('add-online-user', (recipientUserId) => {
+      dispatchConversations({ type: 'ADD_ONLINE_USER', recipientUserId });
+    });
+
+    socket.on('remove-offline-user', (recipientUserId) => {
+      dispatchConversations({ type: 'REMOVE_ONLINE_USER', recipientUserId });
```
Comment (dproc96) [inline client/src/pages/Messages/index.tsx]: Small thing but can you align the name of the event and action to match?

File: server/socket/index.js [hunk 3]
```diff
@@ -47,25 +47,38 @@ exports.appSocket = (server) => {
...
+    })
+
     // when a user goes logout or goes offline
     socket.on("logout", (id) => {
-      if (onlineUsers.includes(id)) {
-        userIndex = onlineUsers.indexOf(id);
-        onlineUsers.splice(userIndex, 1);
+      if (onlineUsers[id]) {
+        delete onlineUsers[id];
         socket.broadcast.emit("remove-offline-user", id);
```
Comment (dproc96) [inline server/socket/index.js]: (optional) This works fine for now but would not scale well, can you think of a way to only emit this event to relevant users?

**Review by dproc96 — APPROVED** [review]
One very small change request and two optional changes, otherwise this is good to go!

File: client/src/pages/Messages/index.tsx (hunk 2 above)
Comment (g4rry420) [reply client/src/pages/Messages/index.tsx]: Sure, I do that

File: server/socket/index.js (hunk 3 above)
Comment (g4rry420) [reply server/socket/index.js]: Sure, I can emit the socket to only users that are in conversation with the current loggedin user.

File: client/src/pages/Messages/ActiveChat/Input.tsx (hunk 1 above)
Comment (g4rry420) [reply client/src/pages/Messages/ActiveChat/Input.tsx]: Sure, I do that

**Pull request merged**